pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -1299,20 +1343,57 @@ func (p *PriorityQueue) PendingPods() ([]*v1.Pod, string) {
 	return result, fmt.Sprintf(pendingPodsSummary, p.activeQ.Len(), p.podBackoffQ.Len(), len(p.unschedulablePods.podInfoMap))
 }
 
+func (p *PriorityQueue) nominatedPodToInfo(np NominatedPod) *framework.PodInfo {
+	pod := np.ToPod()
+	pInfoLookup := newQueuedPodInfoForLookup(pod)
+
+	obj, exists, _ := p.activeQ.Get(pInfoLookup)
+	if exists {
+		queuedPodInfo := obj.(*framework.QueuedPodInfo)
+		return queuedPodInfo.PodInfo
+	}
+
+	queuedPodInfo := p.unschedulablePods.get(pod)
+	if queuedPodInfo != nil {
+		return queuedPodInfo.PodInfo
+	}
+
+	obj, exists, _ = p.podBackoffQ.Get(pInfoLookup)
+	if exists {
+		queuedPodInfo := obj.(*framework.QueuedPodInfo)
+		return queuedPodInfo.PodInfo
+	}
+
+	return &framework.PodInfo{Pod: pod}
+}
+
+func (p *PriorityQueue) nominatedPodsToInfo(nominatedPods []NominatedPod) []*framework.PodInfo {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	pods := make([]*framework.PodInfo, len(nominatedPods))
+	for i, np := range nominatedPods {
+		pods[i] = p.nominatedPodToInfo(np).DeepCopy()
+	}
+	return pods
+}
+
 // Close closes the priority queue.
 func (p *PriorityQueue) Close() {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 	close(p.stop)
+	p.activeQLock.Lock()
+	// closed field is locked by activeQLock as it is checked in Put() without p.lock set.
 	p.closed = true
+	p.activeQLock.Unlock()
```

Comment:
I'm not sure if we really need to take activeQLock here.
The expected flow in Pop is after we `Broadcast` below, `p.cond.Wait()` ([ref](https://github.com/kubernetes/kubernetes/blob/19958a9816562a28c834341e464b76efe26e17c8/pkg/scheduler/internal/queue/scheduling_queue.go#L928)) is unlocked and it `return` ([ref](https://github.com/kubernetes/kubernetes/blob/19958a9816562a28c834341e464b76efe26e17c8/pkg/scheduler/internal/queue/scheduling_queue.go#L926)).
Any possible problematic scenario if we don't take `activeQLock` here?

Reply:
I think the problematic order of operations might be:
1. Pop() is called
2. `p.closed` is checked and its value is false ([ref](https://github.com/kubernetes/kubernetes/blob/19958a9816562a28c834341e464b76efe26e17c8/pkg/scheduler/internal/queue/scheduling_queue.go#L924))
3. Close() is called
4. `p.closed` is set to true (but without locking on `activeQLock`) ([ref](https://github.com/kubernetes/kubernetes/blob/19958a9816562a28c834341e464b76efe26e17c8/pkg/scheduler/internal/queue/scheduling_queue.go#L1387))
5. `p.cond.Broadcast()` in Close() is called ([ref](https://github.com/kubernetes/kubernetes/blob/19958a9816562a28c834341e464b76efe26e17c8/pkg/scheduler/internal/queue/scheduling_queue.go#L1389))
6. Pop() starts waiting on `p.cond.Wait()`, which will be never released ([ref](https://github.com/kubernetes/kubernetes/blob/19958a9816562a28c834341e464b76efe26e17c8/pkg/scheduler/internal/queue/scheduling_queue.go#L928))

By adding the lock on `activeQLock` in Close(), 4. won't be able to be called between 2. and 6.